pkg/kubernetes/service/v1alpha1/build.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	for key, value := range annotations {
+		b.service.object.Annotations[key] = value
+	}
+	return b
+}
+
+// WithAnnotationsNew sets the Annotations field of Service with provided arguments
+func (b *Builder) WithAnnotationsNew(annotations map[string]string) *Builder {
+	if len(annotations) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing annotations"),
+		)
+		return b
+	}
+
+	// copy of original map
+	newannotations := map[string]string{}
+	for key, value := range annotations {
+		newannotations[key] = value
+	}
+
+	// override
+	b.service.object.Annotations = newannotations
+	return b
+}
+
+// WithLabels merges existing labels if any
+// with the ones that are provided here
+func (b *Builder) WithLabels(labels map[string]string) *Builder {
+	if len(labels) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing labels"),
+		)
+		return b
+	}
+
+	if b.service.object.Labels == nil {
+		return b.WithLabelsNew(labels)
+	}
+
+	for key, value := range labels {
+		b.service.object.Labels[key] = value
+	}
+	return b
+}
+
+// WithLabelsNew resets existing labels if any with
+// ones that are provided here
+func (b *Builder) WithLabelsNew(labels map[string]string) *Builder {
+	if len(labels) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing new labels"),
+		)
+		return b
+	}
+
+	// copy of original map
+	newlbls := map[string]string{}
+	for key, value := range labels {
+		newlbls[key] = value
+	}
+
+	// override
+	b.service.object.Labels = newlbls
+	return b
+}
+
+// WithSelectors merges existing selectors if any
+// with the ones that are provided here
+func (b *Builder) WithSelectors(selectors map[string]string) *Builder {
+	if len(selectors) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing selectors"),
+		)
+		return b
+	}
+
+	if b.service.object.Spec.Selector == nil {
+		return b.WithLabelsNew(selectors)
+	}
+
+	for key, value := range selectors {
+		b.service.object.Spec.Selector[key] = value
+	}
+	return b
+}
+
+// WithSelectorsNew resets existing selectors if any with
+// ones that are provided here
+func (b *Builder) WithSelectorsNew(selectors map[string]string) *Builder {
+	if len(selectors) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing selectors"),
+		)
+		return b
+	}
+
+	// copy of original map
+	newslctrs := map[string]string{}
+	for key, value := range selectors {
+		newslctrs[key] = value
+	}
+
+	// override
+	b.service.object.Spec.Selector = newslctrs
+	return b
+}
+
+// WithPorts sets the Ports field of Service with provided arguments
+func (b *Builder) WithPorts(ports []corev1.ServicePort) *Builder {
+	if len(ports) == 0 {
+		b.errs = append(
+			b.errs,
+			errors.New("failed to build service object: missing ports"),
+		)
+		return b
+	}
+	b.service.object.Spec.Ports = ports
```

Comment:
does Spec.Ports gets a copy (or) is that reference to ports that is passed as argument?